services/wallet/ierc20/ierc20.go

```diff
@@ -220,11 +220,11 @@ func (_IERC20 *IERC20CallerSession) Allowance(owner common.Address, spender comm
 // Solidity: function balanceOf(address account) constant returns(uint256)
 func (_IERC20 *IERC20Caller) BalanceOf(opts *bind.CallOpts, account common.Address) (*hexutil.Big, error) {
 	var (
-		ret0 = new(*hexutil.Big)
+		ret0 = new(*big.Int)
 	)
 	out := ret0
 	err := _IERC20.contract.Call(opts, out, "balanceOf", account)
-	return *ret0, err
+	return (*hexutil.Big)(*ret0), err
```

Comment:
in that case problem won't be solved, because we need to write call response to `hexutil.Big`

Reply:
@Samyoul

Reply:
Hey @flexsurfer , are you sure? When I was experimenting with this code, the assignment code makes a `*big.Int`.

Does my suggested change not give the desired results?

I basically pull out the raw `big.Int` data from the `Call` function and then I type assert the new `big.Int` value to a `hexutil.Big`.

---

If you use the following code

```go
func (_IERC20 *IERC20Caller) BalanceOf(opts *bind.CallOpts, account common.Address) (*hexutil.Big, error) {
	var (
		ret0 = new(hexutil.Big)
	)
	out := &ret0
	err := _IERC20.contract.Call(opts, out, "balanceOf", account)
	return ret0, err
}
```

You get the following error:

`abi: cannot umarshal *big.Int in to hexutil.Big`

What this tells me is that the reflection code in the `Set` functionality parses the incoming contract data as `*big.Int`, so if we need to use a `hexutil.Big` higher up the process type asserting the data would be the best solution.

The only exception, would be if the `Call` function uses the `encoding.TextUnmarshaler` interface, but I can't see the use of `UnmarshalText()` in the accounts package.

Reply:
hey @Samyoul so initial issue in status-react that we receive rounded value , my guess we can't use big.Int because it rounds the value and should use hexutil.Big, but im not sure, so if will use big.Int in Call, result still will be rounded ?

Reply:
Yeah, I read through the bug report. I don't think that the problem is with `big.Int` as `big.Int` shouldn't be rounding at all, it is designed for very big numbers.

`hexutil.Big` is a wrapper for `big.Int` with some additional un/marshalling functionality, so unless the `Call` function uses one of the additional interface implementations there is nothing that `hexutil.Big` would do that would fix the problem.

My suggestion is that you do some experimenting and print out the value of a contract at various points in the `status-go` process flow to see if the value is always rounded, or if the value gets rounded in a function inside `status-go`. Does the `status-go` API endpoint return a rounded value?

Reply:
Currently as far as  I can see, `Big.Int` is used as a number in Javascript :
https://github.com/status-im/status-react/blob/0448edb509b705bb48bfefacef38bbaf2cc114de/src/status_im/wallet/core.cljs#L146

That looks like a problem to me, as we might be hitting `Number.MAX_SAFE_INTEGER`  (9007199254740991 on my browser).

We should have the JSON from the RPC return a string and keep that as a string in javascript (or any non-lossy conversion, byte buffers etc), as any conversion lossy might be hitting the limit if we use numbers.

On ours side we probably want to marshalJSON as strings, safer.

Reply:
In our case it will be now a string of hex bytes https://github.com/ethereum/go-ethereum/blob/9a39c6bcb1fe11a252ac832d995a72e5f2210fbe/common/hexutil/json.go#L39 I believe, which is fine if that's what we want.
If we use math/big int it will be converted to a number, loosing precision when parsed in javascript
https://play.golang.org/p/K6CxuHzhI1S , maybe that's the issue

Reply:
@Samyoul you are right, it works, i fixed tests , could you help with lint please